README.md

```diff
@@ -1 +1,319 @@
+  required: {
+    minmax: minMaxDecoder
+  }
+}) // { minmax: [18, 99] }
+```
+
+`minmax` is now typed as `[number, number]` and not as `number[]`
+
+```ts
+const [firstName, lastName] = D.tuple(D.string, D.string).forceDecode(['Michael', 'Jackson'])
+firstName === 'Michael' // true
+lastName === 'Jackson' // true
+```
+
+#### `D.array`
+
+The array decoder takes another decoder with which it tries to decode a whole JSON array.
+
+```ts
+D.array(D.number).decode([1, 2, 3]) // [1, 2, 3]
+
+D.array(D.number).decode([1, 2, 'not a number']) // null
+```
+
+#### `D.record`
+
+This decoder works the same as the array except that it parses object and returns `Record<string, D>`.
+
+```ts
+const data = {
+  preschoolers: 55,
+  student: 124,
+  employed: 133,
+  unemployed: 128,
+  retired: 67
+}
+
+D.record(D.number).decode(data) // succeeds with data as 'Record<string, number>'
+```
+
+#### `D.keyValuePairs`
+
+Key-value pairs decoder works the same was as [`D.record`](#drecord) but returns an array of tuples.
+
+```ts
+// e.g. with data from previous example
+D.keyValuePairs(D.number).decode(data) // succeeds with data as '[[string, number]]'
+```
+
+#### `D.object`
+
+This is probably the most important (and the most complicated?) decoder. You can decode whole typed objects like this:
+
+```ts
+const person = {
+  name: 'Sarah',
+  age: 25
+}
+
+const personDecoder = D.object({
+  required: {
+    name: D.string,
+    age: D.number
+  },
+  optional: {
+    preferredName: D.string
+  }
+})
+
+personDecoder.decode(person) // succeeds
+```
+
+You pass it an object which has `required` and `optional` object with specified fields. Both `required` and `optional` are optional so if you don't have any optional field you can just omit the `optional` field and vice versa.
+
+> Careful, `null` is not a missing value! Null is an actual value which is supposed to be handled with `D.nullable(…)`
+
+Again, if you want the type of `personDecoder`, you can use `D.Output<…>`
+
+```ts
+type Person = D.Output<typeof personDecoder>
+
+// The above is now equivalent to this interface
+interface Person {
+  name: string
+  age: number
+  preferredName?: string
+}
+```
+
+### _Decoder_`.andThen` & chaining
+
+If the built-in types in JSON aren't enough for you can extend the provided decoders. Let's say you want to decode a `Date` from an ISO string.
+
+```ts
+const dateDecoder = D.string.andThen(date => new Date(date))
+// You can now use it with
+// dateDecoder.decode(…)
+
+// Amazingly, TS is smart enough to allow for this:
+type DecodedDate = D.Output<typeof dateDecoder> // is actually Date! and not a string
+```
+
+> Also, if you throw an error from inside the function, the decoder fails as it would fail with a bad type or anything else!
+
+You can use this for:
+- Transforming to different types
+- Renaming fields in objects
+- Performing stricter checks (e.g. string length) and failing the decoder by throwing an error
+
+## ♻️ Similar projects and differences
```

Comment:
For this section, it would probably be better to include links to these packages, so that people can check them out.